server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -2648,4 +2641,161 @@ public void afterRefresh(boolean didRefresh) throws IOException {
+        private long minRequiredCheckpoint = NO_OPS_PERFORMED;
+
+        synchronized void activateNewEngine(Engine newEngine) {
+            assert activeEngine == null : "engine is activated already";
+            assert pendingEngine == null : "engine is being reset";
+            this.activeEngine = newEngine;
+        }
+
+        Engine getActiveEngineOrNull() {
+            return activeEngine;
+        }
+
+        void makeActiveEngineSearchOnly() throws IOException {
+            Engine closing = null;
+            assert activeEngine != null : "engine is not activated yet";
+            synchronized (this) {
+                if ((this.activeEngine instanceof SearchOnlyEngine) == false) {
+                    final SeqNoStats seqNoStats = this.activeEngine.getSeqNoStats(activeEngine.getLastSyncedGlobalCheckpoint());
+                    closing = this.activeEngine;
+                    this.activeEngine = new SearchOnlyEngine(this.activeEngine.config(), seqNoStats);
+                }
+            }
+            IOUtils.close(closing);
+        }
+
+        synchronized Engine getEngineForResync() {
+            final Engine target = pendingEngine != null ? pendingEngine : activeEngine;
+            if (target == null) {
+                throw new AlreadyClosedException("engine is closed");
+            }
+            return target;
+        }
+
+        synchronized boolean hasPendingEngine() {
+            return pendingEngine != null;
+        }
+
+        synchronized void setPendingEngine(Engine pendingEngine, long minRequiredCheckpoint) {
+            assert this.activeEngine != null : "engine is not activated";
+            assert this.pendingEngine == null : "engine is being reset already";
+            this.pendingEngine = pendingEngine;
+            this.minRequiredCheckpoint = minRequiredCheckpoint;
+        }
+
+        synchronized void adjustMinRequiredCheckpoint(long newMinRequiredCheckpoint) {
+            this.minRequiredCheckpoint = Math.min(minRequiredCheckpoint, newMinRequiredCheckpoint);
+        }
+
+        void closePendingEngine() throws IOException {
+            final Engine closing;
+            synchronized (this) {
+                closing = this.pendingEngine;
+                this.pendingEngine = null;
+            }
+            IOUtils.close(closing);
+        }
+
+        void maybeActivatePendingEngine() throws IOException {
+            Engine closing = null;
+            synchronized (this) {
+                if (pendingEngine != null && pendingEngine.getLocalCheckpoint() >= minRequiredCheckpoint) {
+                    // make sure that all acknowledged writes are visible
+                    pendingEngine.refresh("rollback");
+                    closing = this.activeEngine;
+                    this.activeEngine = this.pendingEngine;
+                    this.pendingEngine = null;
+                }
+            }
+            IOUtils.close(closing);
```

Comment:
How do we ensure that searches are not accessing acquireSearcher on the closed engine and switching to the new engine? Also, is there a test that checks that searches (with preference set to this node) continue to work during this transition.

Reply:
There is a small interval in that callers might acquire searchers from the closed engine and hit AlreadClosedException. We can void this entirely by retrying on "AlreadClosedException" if the accessing engine is different from the current engine. However, I am not sure if we should do it.

There is a test which continuously acquires searchers and makes sure that all acknowledged writes are maintained during the transition. https://github.com/elastic/elasticsearch/pull/32867/files#diff-b268f5fefa5837ece96b957e46f628cbR674 (getShardDocUIDs acquires a searcher and uses that searcher to collect document Ids).

Reply:
> However, I am not sure if we should do it.

why is that? We're building all this machinery to have search availability during the transition, except for this very short moment?

I had the same idea about retrying. An alternative would be to do refcounting for closing the engine, to ensure that we only actually close once all in-flight `acquireSearcher` calls have been completed.

Reply:
Sorry, I did not think this carefully. I thought we have to implement the retry on all methods that we support on the search-only, but I was wrong. We only need to implement retry for "get" and "acquire searcher". These two methods should be simple. Thanks for this great question.